### PR TITLE
tracking PR: takes compatible options as node's createReadStream

### PR DESCRIPTION
**Pull request by dominictarr**
We can make this as fast (or faster) than node's read steam, but it ought to be a drop in replacement for 

``` js
var toPull = require('stream-to-pull-stream')
var fs = require('pull-stream')
toPull.source(fs.createReadStream(file, opts))
```

Don't merge this yet, a few more bits to come.

### CODE REVIEW

**Comment by DamonOehlman** [comment]
Cool Dominic - thanks :)  Looking forward to the rest, will add you as a collaborator in the meantime and provide npm publish access.

**Comment by dominictarr** [comment]
great!

**Comment by dominictarr** [comment]
okay now there is better test coverage for abort. an abort can occur _while_ reading or opening, if this happens, wait until the stream is no longer busy, mark the busy cb as ended, then close to the abort cb.

Early versions of pull-stream restricted closing till the current read was finished, but this didn't work for real time streams, which may wait for an arbitrary length of time. Otherwise sometimes you can't abort real time streams.

Speaking of realtime streams, I'd like to support a `live` option, which would be like `tail -f` but I think that will be another pr.

**Comment by dominictarr** [comment]
If you default to a larger buffer size you can read much faster. node uses 64k buffers.
Change how the buffer is allocated, so that the CPU isn't blocked. queue the read op asap, then allocate the buffer we'll use next. Otherwise, we'd be waiting for the buffer to allocate when we could be reading.

This makes it about 15% faster than node on a large file (say 3gb) use `example/rate.js` and `example/node-rate.js`. There is a bit of variation though, so I won't be happy until we have something that will run the benchmark a bunch of times and draw us a curve.

**Comment by DamonOehlman** [comment]
> There is a bit of variation though, so I won't be happy until we have something that will run the benchmark a bunch of times and draw us a curve.

What tools are people using to do this these days?  I'm happy to take a look at this seeing as you are doing some nice work on making this all work the way it's mean to :)

**Comment by dominictarr** [comment]
@DamonOehlman I don't know yet. I was just gonna start by running the test a bunch of times and then calculating the mean and stdev, then at least you can have a reasonable comparison about relative performance.

**Comment by dominictarr** [comment]
okay I added a really simple bench mark to examples - it just reads a file N times with either this module or with node streams, and calculates the probability that a given "race" was one by pull streams.
for a 13 mb file, pull streams win 90% of the time. with a 3 gb file pull streams win every time.

I think it is because of the time to close the stream, which pull streams do before they give the end signal.

**Comment by dominictarr** [comment]
could test that by doing the benchmark by passing in the `FD` and enabling `autoClose`

**Comment by dominictarr** [comment]
This now supports the `encoding` option. If an encoding is provided, it just pulls it through `pull-utf8-decoder`. (much more elegant than encodings in node's streams!)

**Comment by dominictarr** [comment]
So this is now ready for review

**Comment by DamonOehlman** [comment]
Sweet :) Will take a look today.

File: test/largefile.js [hunk 1]
```diff
@@ -16,3 +16,23 @@ test('large file', function(t) {
...
+
+test('large file as ascii strings', function(t) {
+  var expected = ['JTY', 'AU', '01', '0', '609', 'Australia/Sydney', '2012-02-29', ''];
+
+  t.plan(1);
+
+  pull(
+    file(path.resolve(__dirname, 'assets', 'AU.txt'), {encoding: 'ascii'}),
+    pull.through(function (str) {
+      t.equal(typeof str, 'string')
```
Comment (DamonOehlman) [inline test/largefile.js:29]: Is this test meant be here?  Tests currently failing because assertion count != expected.

**Comment by DamonOehlman** [comment]
Awesome stuff Dominic - thanks heaps for this.  Just clarification on whether that test should be actually checking the `typeof` the data going through the stream.  Then I'll merge it all in and look to publish a `1.0.0` version :)

File: test/largefile.js (hunk 1 above)
Comment (dominictarr) [inline test/largefile.js:29]: yes, this is correct. when you set encoding it outputs strings, so put that check in.

**Comment by dominictarr** [comment]
:

**Comment by dominictarr** [comment]
:+

**Comment by dominictarr** [comment]
:+1:

**Comment by dominictarr** [comment]
oh oops, the partial test is failing but I didn't realize because `npm
test` was only running test/all.js

**Comment by DamonOehlman** [comment]
OK - I've merged to master.  Removed the `test/all.js` file and changed the `npm test` command to use an execution that is the same as your other `pull-` modules.  Feel free to push a fix to master :)

**Comment by dominictarr** [comment]
cool, will do, am unlikely to get to this until next week though.

**Comment by dominictarr** [comment]
okay this is fixed, and I changed the largefile test so that it generates a random file so that npm install doesn't need to download a 12mb test case. published as 0.5.0

**Comment by DamonOehlman** [comment]
very cool :)

**Pull request merged**